unstable/lens/set-filterer.rkt

```diff
@@ -0,0 +1,46 @@
+#lang racket/base
+
+(require racket/contract/base)
+(provide (contract-out
+          [set-filterer-lens
+           (-> predicate/c (lens/c functional-set? functional-set?))]
+          ))
+
+(require lens/base/main
+         racket/list
+         racket/set
+         fancy-app
+         )
+(module+ test
+  (require rackunit))
+
+(define (set-filterer-lens pred)
+  (make-lens
+   (set-filter pred _)
+   (λ (tgt nvw)
+     (unless (andmap pred (set->list nvw))
+       (error 'set-filterer-lens "expected (set/c ~a), given: ~v" (contract-name pred) nvw))
+     (set-union (set-filter-not pred tgt) nvw))))
+
+(define (set-filter pred set)
+  (for/fold ([set set]) ([elem (in-set set)] #:unless (pred elem))
+    (set-remove set elem)))
+
+(define (set-filter-not pred set)
+  (for/fold ([set set]) ([elem (in-set set)] #:when (pred elem))
+    (set-remove set elem)))
+
+(define (functional-set? st)
+  (and (generic-set? st)
+       (set-implements? st 'set-add 'set-remove)))
+
+(module+ test
+  (check-equal? (lens-view (set-filterer-lens number?) '(1 a 2 b c 3 d e))
+                '(1 2 3))
+  (check-equal? (lens-set (set-filterer-lens number?) '(1 a 2 b c 3 d e) '(4 5 6 7))
+                '(7 6 5 4 a b c d e))
```

Comment:
Well, lists are sets, so people will expect them to work on lists as well, but they won't expect the kind of ordering that they would expect with a list version.